For #24338 - Make a reusable list item Composable

MozillaNoah opened the pull request:
Closes #24338 

V2 draft to make the root Composable private with specific use case Composables


### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,305 @@
+                tint = FirefoxTheme.colors.iconPrimary,
+            )
+
+            Spacer(modifier = Modifier.width(16.dp))
```
Probably prefer a horizontal padding in the ListItem row

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,305 @@
+    drawableDescription: String? = null,
+    @DrawableRes icon: Int? = null,
+    iconDescription: String? = null,
+    onIconCLick: (() -> Unit)? = null,
```
```suggestion
    onIconClick: (() -> Unit)? = null,
```

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,342 @@
+ * @param onIconClick Optional lambda for handling clicks on the icon.
+ */
+@Composable
+fun FavIconListItem(
```
```suggestion
fun FaviconListItem(
```
This happens to be a nice example that follows nicely with the Figma component naming. It would be nice if we can align our naming nicely with the Component System. I don't think we need a unique component for Left Favicon + Right Icon for instance, but we can take a more generic naming like Favicon List Item as the core component and know that the Right Icon, Descriptions and such will be optional way to extend a core component.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,342 @@
+
+@Composable
+@Preview(name = "FavIconListItem with an icon and onClicks", uiMode = Configuration.UI_MODE_NIGHT_YES)
+private fun CompleteListItemPreview() {
```
Any particular reason these shouldn't just be named after their component `FaviconListItemPreview`? I feel the the @Preview name does provide a helpful description of the intended preview so I am not quite sure if the actual function name needs to be different from the component. I would typically look for the 1:1 component to preview naming if I was trying to see an example of the @Preview

MozillaNoah replied:
The function name is leftover from V1, so that's my bad for not updating it after giving the custom `@Preview` name. I think that's a fair assessment though. For this, I was testing how it looked using a custom preview name to compare apples to apples, but I agree that the function name should align with the component itself.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,342 @@
+ *
+ * @param label The primary text in the item.
+ * @param description An optional second line of text.
+ * @param onClick Optional lambda for handling clicks.
```
```suggestion
 * @param onClick Called when the user clicks on the item.
```

I have mostly been emulating the kDoc conventions from the Jetpack Compose components. I am mostly looking for what actions and its subject (ex, the list item) that will invoke the callback. 

<img width="719" alt="Screen Shot 2022-04-14 at 11 24 06 AM" src="https://user-images.githubusercontent.com/1190888/163422347-875d9382-16eb-49a4-b138-ea2676fff9e1.png">
<img width="621" alt="Screen Shot 2022-04-14 at 11 24 14 AM" src="https://user-images.githubusercontent.com/1190888/163422349-17be60ec-39af-4d6c-96a7-5bca9bea66bf.png">

MozillaNoah replied:
I think this makes sense.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,342 @@
+ * List item Composable used to display text with an optional second line of text and
+ * an optional interactable icon at the end.
+ *
+ * @param label The primary text in the item.
```
```suggestion
 * @param label The label in the list item.
```
nitpicking a bit on these @param doc since these are our **core** reusable components. I don't think we need to use our internal wording ("primary text") since they are called "Label" in the figma design as well.

MozillaNoah replied:
I agree with this. Also didn't think how calling it "primary text" overlaps with the `PrimaryText' Composable

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,342 @@
+ * an optional interactable icon at the end.
+ *
+ * @param label The primary text in the item.
+ * @param description An optional second line of text.
```
```suggestion
 * @param description An optional description text below the label.
```

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,342 @@
+val ICON_SIZE = 24.dp
+
+/**
+ * List item Composable used to display text with an optional second line of text and
```
```suggestion
 * List item used to display a label with an optional description text and
```
Mostly trying to avoid the internal implementation detail ("Composable") and overall just being very clear semantically.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,342 @@
+ * @param label The primary text in the item.
+ * @param description An optional second line of text.
+ * @param onClick Optional lambda for handling clicks.
+ * @param icon Optional DrawableRes used to display a drawable at the end of the list item.
```
```suggestion
 * @param icon Optional [DrawableRes] used to display a drawable at the end of the list item.
```

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,342 @@
+                size = ICON_SIZE,
+            )
+
+            Spacer(modifier = Modifier.width(16.dp))
```
The before and after `Spacer` shouldn't be necessary since `ListItem` already has a horizontal padding of 16dp. I think we just need a 32dp Spacer here.

<img width="866" alt="Screen Shot 2022-04-14 at 11 40 51 AM" src="https://user-images.githubusercontent.com/1190888/163425267-85f655d0-39f2-4b03-bee4-724405b2007e.png">

MozillaNoah replied:
`ListItem` doesn't actually have any padding on the parent `Row` Composable, only on the internal `Column` used for text.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,342 @@
+ * @param label The primary text in the item.
+ * @param description An optional second line of text.
+ * @param onClick Optional lambda for handling clicks.
+ * @param url The tab's URL.
```
I think the description of this can be better. Can you take a look at the Favicon @param for url?

gabrielluong reviewed (changes requested):
Good first pass on this. I didn't look too deeply to see if it fully matched the design. Just thinking about our naming, conventions and how we can make this fit with the Design System.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,342 @@
+    description: String? = null,
+    onClick: (() -> Unit)? = null,
+    url: String,
+    @DrawableRes icon: Int? = null,
+    iconDescription: String? = null,
+    onIconClick: (() -> Unit)? = null,
```
I understand where you're coming from. The rigidity here is because the figma for list items has ONLY an Icon at the end currently, but it's not to say that requirement/spec will never change during the lifetime of the component.

MozillaNoah commented:
Most recent commit updated the kdocs and previews based on suggestions and removed unnecessary padding from right icons.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,338 @@
+        description = description,
+        onClick = onClick,
+        beforeListAction = {
+            Spacer(modifier = Modifier.width(16.dp))
+
+            Favicon(
+                url = url,
+                size = ICON_SIZE,
+            )
+
+            Spacer(modifier = Modifier.width(16.dp))
```
This might be a good instance where we should be adding a Modifier parameter to `Favicon`. and prefer to add a `Modifier.padding(horizontal = 16.dp)` over `Spacer`. Also, see https://chris.banes.dev/always-provide-a-modifier/

MozillaNoah replied:
I did find it odd that `Favicon` wasn't written to have a `Modifier` as a param. I'll make a follow-up story for this.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,338 @@
+        contentAlignment = Alignment.Center,
+    ) {
+        Icon(
+            painter = painterResource(icon),
```
I'll take a swing at this

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,338 @@
+
+        Column(
+            modifier = Modifier
+                .padding(horizontal = 16.dp, vertical = 10.dp)
```
`10.dp` seems a bit arbitrary for such a scenario. I would even prefer to just specify `6.dp` since that was the minimum padding for the label + description scenario.

MozillaNoah replied:
That sounds good to me!

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,338 @@
+            description?.let {
+                SecondaryText(
+                    text = description,
+                    fontSize = 12.sp,
```
Just rechecked this based on padding comment since I was wondering if the label + description was actually getting 10.dp padding vertically since it looked a bit big compare to 6.dp. This should be 14.sp

MozillaNoah replied:
Thanks for catching this

MozillaNoah commented:
Most recent push updates the drawable res params to `Painter`s, renames all of their associated params to `endIconPainter`, `endIconDescription`, `endIconOnClick`, `startIconPainter`, etc, updated the description text size to 14, and changed the fallback vertical padding for the text to 6.dp

MozillaNoah commented:
Created https://github.com/mozilla-mobile/fenix/issues/24873 for the `Favicon` `Modifier`

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,338 @@
+ * @param label The label in the list item.
+ * @param description An optional description text below the label.
+ * @param onClick Called when the user clicks on the item.
+ * @param endIconPainter [Painter] used to display a [ListItemIcon] at the end of the list item.
+ * @param endIconDescription Content description of the end icon.
+ * @param onEndIconClick Called when the user clicks on the end icon.
```
I think we should just drop these `end` prefix. I think it's generally understood based on the top level kDoc that the icon goes to the end.

MozillaNoah replied:
Agreed. Updated

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,338 @@
+    label: String,
+    description: String? = null,
+    onClick: (() -> Unit)? = null,
+    startIconPainter: Painter,
+    startIconDescription: String? = null,
+    endIconPainter: Painter? = null,
+    endIconDescription: String? = null,
+    onEndIconClick: (() -> Unit)? = null,
```
We are kinda mixing start/end with before/after here. I think we should just start with before/after.

MozillaNoah replied:
Agreed. Updated

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt:
```diff
@@ -0,0 +1,338 @@
+    label: String,
+    description: String? = null,
+    onClick: (() -> Unit)? = null,
+    afterIconPainter: Painter? = null,
+    afterIconDescription: String? = null,
+    onAfterIconClick: (() -> Unit)? = null,
```
I am generally in favor of not having any prefix ("after") for TextListItem and FaviconListItem. Keep it simple in these cases where we don't need to decide where the icon is going.

MozillaNoah replied:
I think this makes sense, since the kdoc will indicate where the icon will live in these list items. Updated with those suggestions.
